ports/raspberrypi/common-hal/rp2pio/StateMachine.c

```diff
@@ -1036,123 +1060,331 @@ uint8_t rp2pio_statemachine_program_offset(rp2pio_statemachine_obj_t *self) {
+                self->loop2_read_buf_info = self->loop_read_buf_info;
+
+            }
+        }
+    }
+
+    if (SM_DMA_ALLOCATED_READ(pio_index, sm)) {
+        if (stride_in_bytes != self->background_stride_in_bytes) {
+            mp_raise_ValueError(MP_ERROR_TEXT("Mismatched data size"));
+        }
+        if (swap != self->byteswap) {
+            mp_raise_ValueError(MP_ERROR_TEXT("Mismatched swap flag"));
+        }
+
+        while (self->pending_buffers_read) {
+            RUN_BACKGROUND_TASKS;
+            if (self->user_interruptible && mp_hal_is_interrupted()) {
+                return false;
+            }
+        }
+
+        common_hal_mcu_disable_interrupts();
+        self->next_read_buf_1 = self->once_read_buf_info;
+        self->next_read_buf_2 = self->loop_read_buf_info;
+        self->next_read_buf_3 = self->loop2_read_buf_info;
+        self->pending_buffers_read = pending_buffers_read;
+
+        if (self->dma_completed_read && self->next_read_buf_1.info.len) {
+            rp2pio_statemachine_dma_complete_read(self, SM_DMA_GET_CHANNEL_READ(pio_index, sm));
+            self->dma_completed_read = false;
+        }
+
+        common_hal_mcu_enable_interrupts();
+
+        return true;
+    }
+
+    int channel_read = dma_claim_unused_channel(false);
+    if (channel_read == -1) {
+        return false;
+    }
+    SM_DMA_SET_CHANNEL_READ(pio_index, sm, channel_read);
+
+    // set up receive DMA
+
+    volatile uint8_t *rx_source = (volatile uint8_t *)&self->pio->rxf[self->state_machine];
+
+    self->rx_dreq = pio_get_dreq(self->pio, self->state_machine, false);
+
+    dma_channel_config c_read;
+
+    self->current_read_buf = self->once_read_buf_info;
+    self->next_read_buf_1 = self->loop_read_buf_info;
+    self->next_read_buf_2 = self->loop2_read_buf_info;
+    self->next_read_buf_3 = self->loop_read_buf_info;
+
+    self->pending_buffers_read = pending_buffers_read;
+    self->dma_completed_read = false;
+
+    self->background_stride_in_bytes = stride_in_bytes;
+    self->byteswap = swap;
+
+    c_read = dma_channel_get_default_config(channel_read);
+    channel_config_set_transfer_data_size(&c_read, _stride_to_dma_size(stride_in_bytes));
+    channel_config_set_dreq(&c_read, self->rx_dreq);
+    channel_config_set_read_increment(&c_read, false);
+    channel_config_set_write_increment(&c_read, true);
+    channel_config_set_bswap(&c_read, swap);
+    dma_channel_configure(channel_read, &c_read,
+        self->once_read_buf_info.info.buf,
+        rx_source,
+        self->once_read_buf_info.info.len / stride_in_bytes,
+        false);
+
+    common_hal_mcu_disable_interrupts();
+    // Acknowledge any previous pending interrupt
+    dma_hw->ints0 |= 1u << channel_read;
+    MP_STATE_PORT(background_pio)[channel_read] = self;
+    dma_hw->inte1 |= 1u << channel_read;
```

Comment:
should the assignment to `ints` be `ints1` to match `inte1`?

Reply:
> should the assignment to `ints` be `ints1` to match `inte1`?

No doubt.  Thanks for the review @jepler!